### PR TITLE
implement parallel mult for computing exponent

### PR DESCRIPTION
**Pull request by kwantam**
I had some time on my hands, so I implemented your suggestion for parallelizing multiplication.

Not totally clear to me whether this is the optimal way to do it (e.g., one could imagine doing something more like Strassen's method), but I'm hopeful that it'll give a bit of a speed-up.

### CODE REVIEW

**Comment by alex-ozdemir** [comment]
On robocop this is actually slower :(. I'll have to look into why.

Benchmarked with the `rsa_set_init` example.

File: src/set/int_set_par.rs [hunk 1]
```diff
@@ -391,22 +391,108 @@ where
...
+    use gmp_mpfr_sys::gmp::mpz_swap;
+    use rayon::prelude::*;
+
+    // make sure a is the larger of the two values --- gives smaller operands to muls below
+    if b.significant_bits() > a.significant_bits() {
+        unsafe {
+            let a_ptr = a.as_raw_mut();
+            let b_ptr = b.as_raw_mut();
+            mpz_swap(a_ptr, b_ptr);
+        };
```
Comment (alex-ozdemir) [inline src/set/int_set_par.rs]: Would [`std::mem::swap`](https://doc.rust-lang.org/stable/std/mem/fn.swap.html) suffice for this?

File: src/set/int_set_par.rs [hunk 2]
```diff
@@ -391,22 +391,108 @@ where
...
+
+    // add up the result
+    _parallel_sum(&mut tmp);
+
+    // swap a and tmp[0] ; faster than assigning
+    unsafe {
+        let a_ptr = a.as_raw_mut();
+        let t_ptr = tmp[0].as_raw_mut();
+        mpz_swap(a_ptr, t_ptr);
+    };
```
Comment (alex-ozdemir) [inline src/set/int_set_par.rs]: Would [`std::mem::swap`](https://doc.rust-lang.org/stable/std/mem/fn.swap.html) suffice for this?

**Review by alex-ozdemir — COMMENTED** [review]
Just a pair of FYIs

**Comment by kwantam** [comment]
Hm, yeah, I kind of worried that it'd get slower (didn't know about that benchmark script; thanks!).

I think the issue is that this is a pretty lazy implementation: it only breaks up the left operand, not both. I guess what's going on is, the cost of all the copying to generate the inputs to the parallel mults overwhelms the meager savings. (Might there be a way of reducing that copying? Hmm...)

So I guess Strassen's is the way to go. I'll turn it over in the back of my head, but I'm probably not going to spend more time on this in the immediate future, because I'm guessing that getting this to parallelize nicely is nontrivial and will end up being a rabbit hole.

File: src/set/int_set_par.rs (hunk 1 above)
Comment (kwantam) [reply src/set/int_set_par.rs]: Yeah, probably.

**Comment by alex-ozdemir** [comment]
> I'm probably not going to spend more time on this in the immediate future.

Sounds good.

**Comment by kwantam** [comment]
Thinking about this more, we could probably get some speedup from a Strassen-ish approach. Doesn't have to be perfect, and maybe wouldn't be *too* hard. Maybe I'll try it on the plane.

But fundamentally the way to do this is to make GMP do multi-threaded multiplication. It doesn't do this because it's freakin hard. So maybe we should be OK with a hack, even if it's a crappy one.

**Comment by kwantam** [comment]
OK, fine, couldn't stop myself. I implemented a Strassen-ish multiplication where we split both operands into pieces and compute all the cross terms in parallel.

I played around with trying to be smart about combining the cross terms, but in the end the fastest thing was just to grind them all out, shift them individually, and add in parallel.

On my machine this gives a speedup. Right now parallel_product invokes parallel_multiply only for the final product computation, but I suspect the right heuristic has to do with the size of the operands. See line 169 of parallel_product.rs.

**Comment by kwantam** [comment]
Just tried on tachyon, which has 16 cores / 32 threads, and it's fastest to parallelize the last *two* iterations (i.e., `if split_point < 3` on line 169 of parallel_product.rs). I suspect that either 3 or 5 will be the right answer for robocop.

--

EDIT: oops, that's for 2^18 elements, not 2^20. Might be worthwhile to parallelize more with bigger products.

--

EDIT 2: for 2^20 elements, split_point < 5 is best. On tachyon, no parallelization gives 21 seconds, split_point < 5 gives 16 seconds. The optimality point is pretty flat; 3 and 5 are very similar.

**Comment by alex-ozdemir** [comment]
Nice!
If only all "work avoidance" could be so cool and productive :smile: .

It seems that there is no safe interface for borrowing digits...

**Pull request merged**